fe/fe-core/src/main/java/com/starrocks/scheduler/TaskManager.java

```diff
@@ -0,0 +1,175 @@
+
+package com.starrocks.scheduler;
+
+
+import com.clearspring.analytics.util.Lists;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Maps;
+import com.starrocks.catalog.Column;
+import com.starrocks.catalog.ScalarType;
+import com.starrocks.common.DdlException;
+import com.starrocks.common.util.QueryableReentrantLock;
+import com.starrocks.common.util.Util;
+import com.starrocks.qe.ConnectContext;
+import com.starrocks.qe.ShowResultSet;
+import com.starrocks.qe.ShowResultSetMetaData;
+import com.starrocks.sql.ast.SubmitTaskStmt;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+public class TaskManager {
+
+    private static final Logger LOG = LogManager.getLogger(TaskManager.class);
+
+    public static final long TASK_EXISTS = -1L;
+    public static final long DUPLICATE_CREATE_TASK = -2L;
+    public static final long TASK_CREATE_TIMEOUT = -3L;
+
+    // taskId -> Task , Task may have Manual Task, Periodical Task
+    // every TaskRun must be generated by a Task
+    private final Map<Long, Task> manualTaskMap;
+    // taskName -> Task, include Manual Task, Periodical Task
+    private final Map<String, Task> nameToTaskMap;
+
+    // include PENDING/RUNNING taskRun;
+    private final TaskRunManager taskRunManager;
+
+    // The dispatchTaskScheduler is responsible for periodically checking whether the running TaskRun is completed
+    // and updating the status. It is also responsible for placing pending TaskRun in the running TaskRun queue.
+    // This operation need to consider concurrency.
+    // This scheduler can use notify/wait to optimize later.
+    private final ScheduledExecutorService dispatchScheduler = Executors.newScheduledThreadPool(1);
+
+    // Use to concurrency control
+    private final QueryableReentrantLock lock;
+
+    public TaskManager() {
+
+        manualTaskMap = Maps.newConcurrentMap();
+        nameToTaskMap = Maps.newConcurrentMap();
+        taskRunManager = new TaskRunManager();
+        lock = new QueryableReentrantLock(true);
+
+        dispatchScheduler.scheduleAtFixedRate(() -> {
+            if (!tryLock()) {
+                return;
+            }
+            try {
+                taskRunManager.checkRunningTaskRun();
+                taskRunManager.scheduledPendingTaskRun();
+            } catch (Exception ex) {
+                LOG.warn("failed to dispatch job.", ex);
+            } finally {
+                unlock();
+            }
+
+        }, 0, 1, TimeUnit.SECONDS);
+    }
+
+    public long createTask(Task task) {
+        if (!tryLock()) {
+            return TASK_CREATE_TIMEOUT;
+        }
+        try {
+            if (nameToTaskMap.containsKey(task.getName())) {
+                return TASK_EXISTS;
```

Comment:
what is the difference between DUPLICATE_CREATE_TASK and TAKS_EXISTS?

Reply:
DUPLICATE_CREATE_TASK means that the same TaskID has been submitted twice, which is usually unlikely. It is usually an error of its own system, and TAKS_EXISTS the repetition of the Task Name usually means that the user has submitted the same TASK twice.